src/ZVal.h

```diff
@@ -0,0 +1,658 @@
+	 * @return  True if the value is nil given that type, false otherwise.
+	 */
+	bool IsNil(const zeek::TypePtr& t) const;
+
+	/**
+	 * Return a Val object corresponding to this ZAMValUnion.  If
+	 * the value is managed, then the result is the same underlying
+	 * value.  If not managed, then the result is newly constructed.
+	 * @param t  the type to use in interpreting the ZAMValUnion.
+	 * @return  a ValPtr reflecting the value.
+	 */
+	zeek::ValPtr ToVal(const zeek::TypePtr& t) const;
+
+	// Used for bool, int, enum.
+	bro_int_t int_val;
+
+	// Used for count, counter, port.
+	bro_uint_t uint_val;
+
+	// Used for double, time, interval.
+	double double_val;
+
+	// The types are all variants of Val, zeek::Type, or more fundamentally
+	// zeek::Obj.  They are raw pointers rather than IntrusivePtr's because
+	// unions can't contain the latter.  For memory management, we use
+	// Ref/Unref.
+	zeek::StringVal* string_val;
+	zeek::AddrVal* addr_val;
+	zeek::SubNetVal* subnet_val;
+	zeek::File* file_val;
+	zeek::Func* func_val;
+	zeek::ListVal* list_val;
+	zeek::OpaqueVal* opaque_val;
+	zeek::PatternVal* re_val;
+	zeek::TableVal* table_val;
+	zeek::RecordVal* record_val;
+	zeek::VectorVal* vector_val;
+	zeek::Type* type_val;
+
+	// Used for direct "any" values.
+	zeek::Val* any_val;
+
+	// Used for the compiler to hold opaque items.  Memory management
+	// is explicit in the operations accessing it.
+	val_vec* vvec;
+
+	// Used by the compiler for managing "for" loops.  Implicit
+	// memory management.
+	IterInfo* iter_info;
+
+	// Used for generic access to managed (reference-counted) objects.
+	zeek::Obj* managed_val;
+};
+
+/**
+ * Tests whether a given type is one for which we manage the associated
+ * memory internally.
+ * @param t  the type we want to test for management.
+ * @return  true if the type is managed, false otherwise.
+ */
+bool IsManagedType(const zeek::TypePtr& t);
+
+/**
+ * Deletes a managed value.
+ * @param v  The ZAMValUnion who's managed value should be deleted.
+ */
+inline void DeleteManagedType(ZAMValUnion& v)
+	{
+	Unref(v.managed_val);
+	}
+
+// The following can be set to point to a boolean that will be set
+// to true if a run-time error associated with ZAMValUnion's occurs.
+//
+// We use this somewhat clunky coupling to enable isolating ZVal from
+// ZAM compiler specifics.
+inline bool* zval_error_addr = nullptr;
+
+typedef std::vector<ZAMValUnion> ZVU_vec;
+
+// We need to declare here the external functions that need low-level
+// ZAM_vector access.
+namespace BifFunc {
+extern zeek::detail::BifReturnVal
+		sort_bif(zeek::detail::Frame* frame, const zeek::Args*);
+}
+
+class ZAM_vector {
+public:
+	// In the following, we use a bare pointer for the VectorVal
+	// due to tricky memory management concerns, namely that ZAM_vector's
+	// point to their VectorVal's and VectorVal's point to their
+	// ZAM_vector's.
+	ZAM_vector(zeek::VectorVal* _vv, zeek::TypePtr yt, int n = 0)
+	: zvec(n)
+		{
+		vv = _vv;
```

Comment:
what is vv for? it's a protected member that doesn't seem to be used anywhere?

(and: why are all the members protected instead of private?)

Reply:
Re `vv`, good catch - it's vestigial, as is `rv` in `ZAM_record`.  I'll remove them if we wind up keeping these classes.

Re why protected rather than private: just because that's the style still used in large parts of the code base.  Are we switching to a style of using private in cases like this?

Reply:
> Re why protected rather than private: just because that's the style still used in large parts of the code base.

Ah, I guess never really realized that. I think most newer classes use `private`, and `protected` only for stuff that indeed needs access from derived classes. I would suggest we use that approach going forward as well in the interest of encapsulation, but point taken that that's not settled style.